packages/react-router/lib/dom/ssr/routes.tsx

```diff
@@ -64,7 +69,22 @@ function getRouteComponents(
   routeModule: RouteModule,
   isSpaMode: boolean
 ) {
-  let Component = getRouteModuleComponent(routeModule);
+  let ComponentNeedingProps = getRouteModuleComponent(routeModule);
+  let Component = ComponentNeedingProps
+    ? () => {
+        let params = useParams();
+        let data = useLoaderData();
+        let actionData = useActionData();
+        return (
+          <ComponentNeedingProps
+            // @ts-expect-error
+            params={params}
+            data={data}
+            actionData={actionData}
+          />
+        );
+      }
+    : undefined;
```

Comment:
right now HMR tests are failing. pretty sure that this part is to blame since were dynamically creating a new component every time we want to render the route's component.

for a quick implementation, I think it'd be easiest to just create a virtual mod for each route (similar to what we already do for production) that re-exports from the route module's default export and wraps the `Component`.

a more principled solution would be to find out how we pass in data from loaders and actions and stop relying on hooks altogether. that would probably remove the need to wrap the `Component` from a route.